src/Servant/Reflex/Response.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE ConstraintKinds   #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Response handlers
+--   Usefull for 'clientWithOptsAndResultHandler'
+module Servant.Reflex.Response
+  ( reloadOnAPIError
+  , reloadOnResponse
+  , RefreshSupport
+  , isAuthErr
+  , is4xxErr
+  , isDecodeError
+  ) where
+
+import           Control.Monad               (join, void)
+import qualified Data.Text                   as Text
+import           JSDOM                       (currentDocument)
+import           JSDOM.Generated.Document    (getLocation)
+import           JSDOM.Types                 (Location, liftDOM)
+import           Language.Javascript.JSaddle (MonadJSM, ( # ))
+import           Reflex
+import           Reflex.Dom.Builder.Class
+import           Reflex.Dom.Widget.Basic
+import           Reflex.Dom.Xhr
+import           Servant.Reflex
+
+-- | Alias for constraints that can refresh
+type RefreshSupport t m
+   = (MonadJSM m, MonadHold t m, DomBuilder t m, SupportsServantReflex t m)
+
+is4xxErr :: ReqResult () a -> Bool
+is4xxErr =
+  maybe
+    False
+    (\req ->
+       (_xhrResponse_status req < fromInteger 500) &&
+       (_xhrResponse_status req >= 400)) .
+  response
+
+isAuthErr :: ReqResult () a -> Bool
+isAuthErr =
+  maybe
+    False
+    (\req ->
+       let stat = _xhrResponse_status req
+       in (stat == fromInteger 401) &&
+           (stat == fromInteger 402) && (stat == fromInteger 407)) .
+  response
+
+isDecodeError :: ReqResult () a -> Bool -- TODO find a better way of detecting this, is there no typesafe decode error?
+isDecodeError = maybe False (Text.isInfixOf "Error in $") . reqFailure
+
+-- | Reloads page and clears the cache on any 4xx response which is not auth related
+--   or on any decoding errors.
+--   Most endpoints want to have this.
```

Comment:
I can imagine CRUD related 404s that would result in continuously reloading the page. In your experience using this function, have you ever seen it? Was it hard to debug (since you wouldn't be able to get a dev console up?)

Reply:
I've had reload loops because of this code. Because this is the only code that does reloading I didn't really had to do any kind of debugging. In my case I just isolated the endpoints that expected 404.

Since it takes a little while for relfex to 'boot' (half a sec on chrome, longer on ff). there is some time to open up the console. If you preserve logs you can see what goes on regardless of reloads.